src/Service/Services/ExecutionHelper.cs

```diff
@@ -0,0 +1,431 @@
+                    ByteArrayType => fieldValue.GetBytesFromBase64(),
+                    BooleanType => fieldValue.GetBoolean(), // spec
+                    UrlType => new Uri(fieldValue.GetString()!),
+                    UuidType => fieldValue.GetGuid(),
+                    TimeSpanType => TimeSpan.Parse(fieldValue.GetString()!),
+                    _ => fieldValue.GetString()
+                };
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Represents a pure resolver for an object field.
+        /// This resolver extracts another json object from the parent json property.
+        /// </summary>
+        /// <param name="context">
+        /// The pure resolver context.
+        /// </param>
+        /// <returns>
+        /// Returns a new json object.
+        /// </returns>
+        public object? ExecuteObjectField(IPureResolverContext context)
+        {
+            if (TryGetPropertyFromParent(context, out JsonElement objectValue) &&
+                objectValue.ValueKind is not JsonValueKind.Null and not JsonValueKind.Undefined)
+            {
+                IMetadata metadata = GetMetadata(context);
+                objectValue = _queryEngine.ResolveInnerObject(objectValue, context.Selection.Field, ref metadata);
+                
+                // Since the query engine could null the object out we need to check again
+                // if its null.
+                if(objectValue.ValueKind is not JsonValueKind.Null and not JsonValueKind.Undefined)
+                {
+                    return null;
+                }
+                
+                SetNewMetadata(context, metadata);
+                return objectValue;
+            }
+
+            return null;
+        }
+
+        public object? ExecuteListField(IPureResolverContext context)
+        {
+            if (TryGetPropertyFromParent(context, out JsonElement listValue) && 
+                listValue.ValueKind is not JsonValueKind.Null and not JsonValueKind.Undefined)
+            {
+                IMetadata metadata = GetMetadata(context);
+                IReadOnlyList<JsonElement> result = _queryEngine.ResolveListType(listValue, context.Selection.Field, ref metadata);
+                SetNewMetadata(context, metadata);
+                return result;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Set the context's result and dispose properly. If result is not null
+        /// clone root and dispose, otherwise set to null.
+        /// </summary>
+        /// <param name="context">Context to store result.</param>
+        /// <param name="result">Result to store in context.</param>
+        private static void SetContextResult(IMiddlewareContext context, JsonDocument? result)
+        {
+            if (result is not null)
+            {
+                context.RegisterForCleanup(() => result.Dispose());
+                context.Result = result.RootElement;
```

Comment:
>result.RootElement [](http://example.com/codeflow?start=33&length=18)

Is it not required to copy the JsonElement using `result.RootElement.Clone()` because of the registration of `result.Dispose()` to `context.RegisterForCleanup()`? 
I see in HC IMiddlewareContext.cs comment:  
> "Register cleanup tasks that will be executed after resolver execution is finished."

So essentially confirming my understanding here that this is okay, because the JsonDocument won't be disposed until the resolver is done executing?

Reply:
Since the lower level pure resolvers are folded into the execution of the async resolver disposing it after the resolver is finished is good enough. In version 13 we have introduced an option to dispose when the result is disposed.

`void RegisterForCleanup(Func<ValueTask> action, CleanAfter cleanAfter = CleanAfter.Resolver);`

https://github.com/ChilliCream/graphql-platform/blob/820384f9ddc7104b8f34cde6872e60326b93be8b/src/HotChocolate/Core/src/Types/Resolvers/IMiddlewareContext.cs#L57

We can do this also in version 13 but it takes more effort.